base/tuple.jl

```diff
@@ -73,14 +73,15 @@ keys(@nospecialize t::Tuple) = OneTo(length(t))
 prevind(@nospecialize(t::Tuple), i::Integer) = Int(i)-1
 nextind(@nospecialize(t::Tuple), i::Integer) = Int(i)+1
 
-function keys(t::Tuple, t2::Tuple...)
-    @inline
-    OneTo(_maxlength(t, t2...))
+@noinline function throw_eachindex_mismatch_indices(::Tuple{}, inds...)
+    throw(DimensionMismatch("all inputs to eachindex must have the same length, got $(join(inds, ", ", " and "))"))
```

Comment:
```suggestion
function throw_eachindex_mismatch_indices(::Tuple{}, inds...)
    @noinline 
    throw(DimensionMismatch("all inputs to eachindex must have the same length, got $(join(inds, ", ", " and "))"))
```

Looks like we have a bootstrapping problem: `tuple` is used very early in Core.Compiler and the full Julia machinery is not yet available. Moving this inside the method definition should resolve the problem, I think

Reply:
@timholy I think we can use `IndexLinear` dispatch here. I do not understand the bootstrapping problem, but I think we will probably run into it again. The function signature in `base/abstractarray.jl` is
https://github.com/JuliaLang/julia/blob/883c19bdcfdec6d4717bce6ba2e652dc8efb10c4/base/abstractarray.jl#L321-L323

Do you suggest we keep the dispatch over `::Tuple{}` with `@noinline` inside?

Reply:
Bootstrapping problems are a pain and trap even very experienced developers. A way to think about it is that much of Julia is written in Julia, starting from a small set of primitives and building up additional features. `base/tuple.jl` gets loaded very early on (tuples are quite fundamental) and presumably provides functionality needed to support complex macros. For an "internal" `@noinline` with no arguments we can implement that with very little functionality (see https://github.com/JuliaLang/julia/blob/883c19bdcfdec6d4717bce6ba2e652dc8efb10c4/base/Base.jl#L26), but the version that takes the function as an argument requires more functionality to implement (see https://github.com/JuliaLang/julia/blob/883c19bdcfdec6d4717bce6ba2e652dc8efb10c4/base/expr.jl#L896-L909); in particular, we need to have defined arrays with operations like `push!`, etc, and we aren't yet there in the process of "bootstrapping" Julia. Does that make more sense?

For the question of whether we could use the LinearIndices version, I'm not entirely certain, but I think it will be safe. The good news is that this only needs to be parsed & lowered early (macros are expanded during lowering), but not yet compiled. So it's OK to define methods that call other methods that haven't yet been defined, as long as we don't actually trigger that call until later. I'd try dispatching to that `LinearIndices` version (it's not really even a lie...) and seeing if it works.

Reply:
You also have to delete/comment-out the version you added. The problem is that `@noinline` in front of the function definition runs when the file gets loaded, regardless of whether you ever call it.

Reply:
Yes, I kept it around for testing. Builds are passing  with your suggestion to move `@noninline` inside.  With `IndexLinear()`, the builds fail again, because of the same reason.
https://github.com/JuliaLang/julia/blob/30a73de68108b9629c4c99a1751d579ef5893bda/base/abstractarray.jl#L321-L323

Reply:
> You also have to delete/comment-out the version you added. The problem is that `@noinline` in front of the function definition runs when the file gets loaded, regardless of whether you ever call it.
No wait, it could be happening because I kept the original version of the function for ::Tuple{}. I will check.